File: php/balanced_brackets.php
<?php

function markers() {
    return array(
        '{' => '}',
        '(' => ')',
        '[' => ']'
    );
}
$buffer = array();
function right($str, $marker) {
    $markers = markers();
    if(empty($str)) return null;
    $c = substr($str, 0, 1); $str = substr($str,1);
    if($c === $marker) return $str;                               //first character is the bracket we wanted.
    if(in_array($c, array_values($markers), true)) return null;   //1st character is a closing bracket, but not the one we wanted.

    if($markers[$c]) {                                            //found another left bracket,
        $str = right($str, $markers[$c]);                         //initiate inner round search
        return ($str === null) ? null                             //inner round search failed
                               : right($str, $marker);            //succeeded - continue outter round match.
    }
    //other character...keep looking
    return right($str, $marker);
}
function braces($str) {
    $markers = markers();
    if(empty($str)) return 'YES';                                  //empty ok.
    $c = substr($str, 0, 1); $str = substr($str,1);

    if(in_array($c, array_values($markers))) return 'NO';          //found right bracket, bail out.
    if($markers[$c]) {                                             // left bracket, starting looking for closing bracket.
        $str = right($str, $markers[$c]);
        if(null === $str) return 'NO';                             //search failed.
    }
    return braces($str);                                           //other character...keep looking
}

$handle = fopen ("php://stdin","r");
fscanf($handle,"%d",$t);
for($a0 = 0; $a0 < $t; $a0++){
    fscanf($handle,"%s",$s);
    print braces($s) . "\n";
}

?>

File: erlang/balanced_brackets.erl

% is A left brace?
is_left(A) ->
  case A of
    ${ -> { yes, $} };
    $( -> { yes, $) };
    $[ -> { yes, $] };
    _  -> { no, A }
  end.

% is A right brace?
is_right(A) ->
  case A of
    $} -> { yes, ${ };
    $) -> { yes, $( };
    $] -> { yes, $[ };
    _  -> { no, A }
  end.

% right: find right brace
% right brace is the first character
right([H|T], H) ->  {ok, T};
% never find the right brace
right([], _) -> { nok, [] };

right([H|T], C) ->
  case is_right(H) of
    { no, _ } ->
      case is_left(H) of
        { no, _ } -> right(T, C);             % either left nor right brace, keep looking
        { yes, C2 } ->                        % another left brace - start another round of looking for closing brace
          case right(T, C2) of
            {ok, T1 }      -> right(T1, C);   % inner round matched; look for outter closing brace
            {nok, T1 }     -> { nok, T1 }     % inner match failed.
          end
      end;
    { yes, _ } -> {nok, T }                   % although H is closing brace, it's not the one we're looking for.
  end.


braces([H|T]) ->
  case is_right(H) of
    { no, _ } ->
      case is_left(H) of
        { no, _ }  -> braces(T);              % neither  left nor right, keep looking
        { yes, C } -> case right(T, C) of     % initiate closing brace match
          { ok, T1 } -> braces(T1);           % matched, move to next set
          { nok, _ } -> "NO"
          end
      end;
    { yes, _ } -> "NO"                        % got right brace off the bat.
  end;

braces([]) -> "YES".



main() ->

    { ok, [T]} = io:fread("", "~d"),
    T_list = lists:seq(1,T),
    lists:foreach( fun(T0) ->
            { ok, [S]} = io:fread("", "~s"),

             io:format("~s~n", [braces(S)] )

    end,T_list),
    true.
